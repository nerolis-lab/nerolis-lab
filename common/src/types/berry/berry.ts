export interface Berry {
  name: string;
  value: number;
  type: string;
}
export interface BerrySet {
  amount: number;
  berry: Berry;
  level: number;
}
export interface BerrySetSimple {
  amount: number;
  name: string;
  level: number;
}
export type BerryIndexToIntAmount = Int16Array;
export type BerryIndexToFloatAmount = Float32Array;
