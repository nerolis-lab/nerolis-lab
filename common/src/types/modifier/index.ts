export * from './modifier';
export * from './target-types';
export * from './path-reference';
