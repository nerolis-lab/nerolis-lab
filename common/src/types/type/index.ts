export * from './type-paths';
export * from './path-value';
