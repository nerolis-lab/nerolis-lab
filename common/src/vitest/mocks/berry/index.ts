export * from './mock-berry';
export * from './mock-berry-set';
