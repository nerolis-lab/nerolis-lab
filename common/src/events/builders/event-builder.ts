import type { Event } from '../../types/events/event-types';
import type { Modifier } from '../../types/modifier/modifier';
import type { ModifierTargetType } from '../../types/modifier/target-types';

/**
 * Function that creates a modifier with access to input
 */
type ModifierFactory<TInput> = (input: TInput) => Modifier<ModifierTargetType>;

class EventBuilderImpl<TInput> {
  private event: Partial<Event> = {
    modifiers: []
  };
  public modifierFactories: ModifierFactory<TInput>[] = [];

  /**
   * Sets the name of the event
   */
  name(name: string): this {
    this.event.name = name;
    return this;
  }

  /**
   * Sets the description of the event
   */
  description(description: string): this {
    this.event.description = description;
    return this;
  }

  /**
   * Sets the start date of the event
   */
  startDate(date: Date): this {
    this.event.startDate = date;
    return this;
  }

  /**
   * Sets the end date of the event
   */
  endDate(date: Date): this {
    this.event.endDate = date;
    return this;
  }

  /**
   * Adds a modifier to the event
   * For static events (TInput = void), only accepts Modifier objects
   * For parameterized events, accepts both Modifier objects and ModifierFactory functions
   */
  addModifier(
    modifier: TInput extends void
      ? Modifier<ModifierTargetType>
      : Modifier<ModifierTargetType> | ModifierFactory<TInput>
  ): this {
    if (typeof modifier === 'function') {
      this.modifierFactories.push(modifier as ModifierFactory<TInput>);
    } else {
      if (!this.event.modifiers) {
        this.event.modifiers = [];
      }
      this.event.modifiers.push(modifier);
    }
    return this;
  }

  /**
   * Builds the event (static events - no input required)
   * If the builder has modifier factories, returns a function (input: TInput) => Event
   * If the builder has no modifier factories, returns Event directly
   */
  build(): TInput extends void ? Event : (input: TInput) => Event;

  /**
   * Builds the event with input for modifier factories (parameterized events)
   */
  build(input: TInput): Event;

  /**
   * Implementation of build method
   */
  build(input?: TInput): Event | ((input: TInput) => Event) {
    if (!this.event.name) {
      throw new Error('Event name is required');
    }

    if (!this.event.description) {
      throw new Error('Event description is required');
    }

    // If we have modifier factories but no input, return a function
    if (this.modifierFactories.length > 0 && input === undefined) {
      return (factoryInput: TInput) => {
        const dynamicModifiers = this.modifierFactories.map((factory) => factory(factoryInput));
        return {
          name: this.event.name!,
          description: this.event.description!,
          modifiers: [...(this.event.modifiers || []), ...dynamicModifiers],
          startDate: this.event.startDate,
          endDate: this.event.endDate
        };
      };
    }

    // Build dynamic modifiers from factories if we have them and input
    const dynamicModifiers =
      this.modifierFactories.length > 0 && input !== undefined
        ? this.modifierFactories.map((factory) => factory(input))
        : [];

    return {
      name: this.event.name,
      description: this.event.description,
      modifiers: [...(this.event.modifiers || []), ...dynamicModifiers],
      startDate: this.event.startDate,
      endDate: this.event.endDate
    };
  }
}

/**
 * Creates an EventBuilder for building static events (no input parameters)
 *
 * // Static event
 * EventBuilder()
 *   .name('test')
 *   .description('test')
 *   .build()
 */
export function EventBuilder(): EventBuilderImpl<void>;

/**
 * Creates an EventBuilder for building parameterized events with typed input
 *
 * @example
 * // Parameterized event
 * const myEventBuilder = EventBuilder<MyInput>()
 *   .name('test')
 *   .description('test')
 *   .addModifier(input => ({
 *     type: 'Pokemon',
 *     path: 'frequency',
 *     operation: '*',
 *     value: input.multiplier
 *   }));
 *
 * // Use: myEventBuilder.build({ multiplier: 0.9 })
 */
export function EventBuilder<TInput>(): EventBuilderImpl<TInput>;

/**
 * Implementation of EventBuilder factory
 * @example
 * // Static event
 * EventBuilder()
 *   .name('test')
 *   .description('test')
 *   .build()
 *
 * // Parameterized event
 * const myEventBuilder = EventBuilder<MyInput>()
 *   .name('test')
 *   .description('test')
 *   .addModifier(input => ({
 *     type: 'Pokemon',
 *     path: 'frequency',
 *     operation: '*',
 *     value: input.multiplier
 *   }));
 *
 * // Use: myEventBuilder.build({ multiplier: 0.9 })
 */
export function EventBuilder<TInput = void>(): EventBuilderImpl<TInput> {
  return new EventBuilderImpl<TInput>();
}
