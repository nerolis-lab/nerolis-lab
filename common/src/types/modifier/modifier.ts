import type { PathKeys, PathValue } from '../type';
import type { PathReference } from './path-reference';
import type { ModifierTargetType, ModifierTargetTypeDTO, TargetTypeMap } from './target-types';

// ============================================
// Core Modifier Types (Internal)
// ============================================

export type ModifierOperation = '*' | '+' | '-' | '/' | '=';
export type ConditionOperation = '=' | '!=' | '>' | '<' | '>=' | '<=' | 'in' | 'not-in';

export type ModifierComplexValue<T extends ModifierTargetType, P extends PathKeys<T>> = {
  rightValue: PathValue<T, P> | PathReference<T>;
  /**
   * If min is provided, the modifier will be at least the min value
   */
  min?: PathValue<T, P> | PathReference<T>;
  /**
   * If max is provided, the modifier will be capped at the max value
   */
  max?: PathValue<T, P> | PathReference<T>;
};

/**
 * Modifier value supports simple values, PathReferences, or constrained operations
 *
 * @example
 * // Simple value
 * 5
 *
 * @example
 * // PathReference to another property
 * PathReference.to('pokemon.skill.maxLevel')
 *
 * @example
 * // Constrained operation (operation comes from top-level modifier)
 * {
 *   rightValue: 3,
 *   max: PathReference.to('pokemon.skill.maxLevel')
 * }
 *
 * @example
 * // Subtract with floor constraint
 * {
 *   rightValue: 1000,
 *   min: 500
 * }
 *
 * @example
 * // Multiply with both min and max constraints
 * {
 *   rightValue: 1.5,
 *   min: 100,
 *   max: 5000
 * }
 */
export type ModifierValue<T extends ModifierTargetType, P extends PathKeys<T>> =
  | PathValue<T, P>
  | PathReference<T>
  | ModifierComplexValue<T, P>;

/**
 * Modifier options that define how to modify a target object
 * Uses consistent leftValue/rightValue naming for clarity:
 * - leftValue: what gets modified (the target path)
 * - operation: how it gets modified
 * - rightValue: what it gets modified by/to
 * - conditions: when the modifier should apply
 */
export type ModifierOptions<T extends ModifierTargetType> = {
  [P in PathKeys<T>]: {
    type: ModifierTargetTypeDTO;
    leftValue: P;
    operation: ModifierOperation;
    rightValue: ModifierValue<T, P>;
    conditions?: Condition<T>[];
  };
};

/**
 * Target condition that checks properties on the target object (Pokemon/PokemonInstance)
 * Uses consistent leftValue/rightValue naming for clarity:
 * - leftValue: what gets checked (the target path)
 * - operation: how it gets compared
 * - rightValue: what it gets compared against
 *
 * RightValue can be either:
 * - A literal value of the expected type
 * - A PathReference to another path on the same object (e.g., PathReference.to('maxSkillLevel'))
 */
type TargetConditionOptions<T extends ModifierTargetType> = {
  [P in PathKeys<T>]: {
    leftValue: P;
    operation: ConditionOperation;
    rightValue: PathValue<T, P> | PathValue<T, P>[] | PathReference<T>;
  };
};

/**
 * External condition that checks external values (input, imports, etc.)
 * Uses consistent leftValue/rightValue naming for clarity:
 * - leftValue: what gets checked (the external value)
 * - operation: how it gets compared
 * - rightValue: what it gets compared against
 * Provides type safety for the value being checked
 */
export type ExternalCondition<TValue = unknown> = {
  leftValue: TValue;
  operation: ConditionOperation;
  rightValue: TValue | TValue[];
};

/**
 * Type-safe modifier interface with optional conditions
 * Uses distributive conditional types to handle union types correctly
 *
 * @example
 * // Modifier that reduces frequency by 10% only for berry specialists
 * const modifier: Modifier<Pokemon> = {
 *   type: 'Pokemon',
 *   path: 'frequency',
 *   operation: '*',
 *   value: 0.9,
 *   condition: {
 *     path: 'specialty',
 *     operation: '=',
 *     value: 'berry'
 *   }
 * };
 */
export type Modifier<TargetType extends ModifierTargetType> = TargetType extends unknown
  ? ModifierOptions<TargetType>[PathKeys<TargetType>]
  : never;
export type ModifierDTO<TargetType extends ModifierTargetTypeDTO = ModifierTargetTypeDTO> = {
  type: TargetType;
  modifiers: Modifier<TargetTypeMap[TargetType]>[];
};
export type TargetCondition<T extends ModifierTargetType> = TargetConditionOptions<T>[PathKeys<T>];
export type Condition<T extends ModifierTargetType> = TargetCondition<T> | ExternalCondition;
