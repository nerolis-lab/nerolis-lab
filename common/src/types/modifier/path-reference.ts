import type { PathKeys } from '../type/type-paths';
import type { ModifierTargetType } from './target-types';

/**
 * Wrapper type to explicitly mark a value as a path reference
 * This allows us to distinguish between literal values and path references at runtime
 *
 * @template T - The target type (Pokemon | PokemonInstanceExt) for type-safe path validation
 */
export class PathReference<T extends ModifierTargetType = ModifierTargetType> {
  constructor(public readonly path: PathKeys<T>) {}

  /**
   * Helper function to create a path reference with type safety
   * @template T - The target type for path validation
   * @param path - Type-safe path from PathKeys<T>
   */
  static to<T extends ModifierTargetType>(path: PathKeys<T>): PathReference<T> {
    return new PathReference<T>(path);
  }
}

/**
 * Type guard to check if a value is a PathReference
 */
export function isPathReference(value: unknown): value is PathReference {
  return value instanceof PathReference;
}

/**
 * Helper function to create a path reference (shorter syntax)
 * @template T - The target type for path validation
 * @param path - Type-safe path from PathKeys<T>
 */
export function pathRef<T extends ModifierTargetType>(path: PathKeys<T>): PathReference<T> {
  return PathReference.to<T>(path);
}
