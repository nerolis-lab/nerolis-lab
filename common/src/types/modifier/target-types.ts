import type { PokemonInstanceExt } from '../instance';
import type { MemberStrength } from '../team';

export type ModifierTargetTypeDTO = 'PokemonInstance' | 'MemberStrength';
export type ModifierTargetType = PokemonInstanceExt | MemberStrength;

/**
 * Used for mapping API strings to actual types
 *
 * @tutorial How to add a new target type
 * 1. Add the string literal to ModifierTargetTypeDTO union above
 * 2. Add the actual type to ModifierTargetType union above
 * 3. Add the mapping entry here
 *
 * Example for adding a new "Item" type:
 * - ModifierTargetTypeDTO = 'Pokemon' | 'PokemonInstanceExt' | 'Item'
 * - ModifierTargetType = Pokemon | PokemonInstanceExt | Item
 * - TargetTypeMap = { Pokemon: Pokemon, PokemonInstanceExt: PokemonInstanceExt, Item: Item } *
 */
export type TargetTypeMap = {
  PokemonInstance: PokemonInstanceExt;
  MemberStrength: MemberStrength;
};
