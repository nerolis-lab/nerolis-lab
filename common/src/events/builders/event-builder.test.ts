import { describe, expect, it } from 'vitest';
import type { ExpertModeSettings } from '../../types';
import { MAGO, ORAN, SITRUS } from '../../types/berry/berries';
import { PathReference } from '../../types/modifier/path-reference';
import { EventBuilder } from './event-builder';

describe('EventBuilder', () => {
  it('should create an event', () => {
    const event = EventBuilder().name('Test Event').description('A test event').build();
    expect(event).toBeDefined();
  });

  it('should create an event with a modifier', () => {
    const event = EventBuilder()
      .name('Test Event')
      .description('A test event')
      .addModifier({ type: 'PokemonInstance', leftValue: 'pokemon.frequency', operation: '*', rightValue: 0.9 })
      .build();
    expect(event).toBeDefined();
  });

  it('should create an event with a modifier with a condition', () => {
    const event = EventBuilder()
      .name('Test Event')
      .description('A test event')
      .addModifier({
        type: 'PokemonInstance',
        leftValue: 'pokemon.frequency',
        operation: '*',
        rightValue: 0.9,
        conditions: [{ leftValue: 'pokemon.specialty', operation: '=', rightValue: 'berry' }]
      })
      .build();
    expect(event).toBeDefined();
  });

  it('should have proper type safety for modifiers', () => {
    const event = EventBuilder()
      .name('Test Event')
      .description('A test event')
      .addModifier({ type: 'PokemonInstance', leftValue: 'pokemon.frequency', operation: '*', rightValue: 0.9 })
      .addModifier({ type: 'PokemonInstance', leftValue: 'pokemon.frequency', operation: '*', rightValue: 0.9 })
      // @ts-expect-error - rightValue should be a number
      .addModifier({ type: 'PokemonInstance', leftValue: 'pokemon.frequency', operation: '*', rightValue: '0.9' })
      // @ts-expect-error - invalid leftValue
      .addModifier({ type: 'PokemonInstance', leftValue: 'pokemon.slkfjklsd', operation: '*', rightValue: 0.9 })
      .addModifier({
        type: 'PokemonInstance',
        leftValue: 'pokemon.frequency',
        operation: '*',
        rightValue: 0.9,
        conditions: [{ leftValue: 'pokemon.berry.name', operation: '=', rightValue: 'MAGO' }]
      })
      .addModifier({
        type: 'PokemonInstance',
        leftValue: 'pokemon.frequency',
        operation: '*',
        rightValue: 0.9,
        conditions: [{ leftValue: 'pokemon.berry.name', operation: '=', rightValue: 'INVALID' }]
      })
      .build();
    expect(event).toBeDefined();
    expect(event.modifiers).toHaveLength(6);
    expect(event.modifiers[0].type).toBe('PokemonInstance');
    expect(event.modifiers[1].type).toBe('PokemonInstance');

    expect(event).toMatchSnapshot();
  });

  it('should create realistic Greengrass Island expert mode identifier', () => {
    const expertModeEvent = EventBuilder<ExpertModeSettings>()
      .name('Expert Mode Event')
      .description('Dynamic event based on input parameters')

      // Main berry gets frequency buff - using modifier factory
      .addModifier((input) => ({
        type: 'PokemonInstance',
        leftValue: 'pokemon.frequency',
        operation: '*',
        rightValue: 0.9,
        conditions: [
          {
            leftValue: 'pokemon.berry.name',
            operation: '=',
            rightValue: input.mainFavoriteBerry.name
          }
        ]
      }))
      .addModifier((input) => ({
        type: 'PokemonInstance',
        leftValue: 'skillLevel',
        operation: '+',
        rightValue: 1,
        conditions: [
          {
            leftValue: 'pokemon.berry.name',
            operation: '=',
            rightValue: input.mainFavoriteBerry.name
          },
          {
            leftValue: 'pokemon.skillLevel',
            operation: '<',
            rightValue: PathReference.to('pokemon.skill.maxLevel')
          }
        ]
      }))
      // random site effect
      .addModifier((input) => ({
        type: 'PokemonInstance',
        leftValue: 'pokemon.skillPercentage',
        operation: '*',
        rightValue: 1.25,
        conditions: [
          {
            leftValue: 'pokemon.berry.name',
            operation: 'in',
            rightValue: [input.mainFavoriteBerry.name, ...input.subFavoriteBerries.map((b) => b.name)]
          }
        ]
      }))
      .build();

    const settings: ExpertModeSettings = {
      mainFavoriteBerry: MAGO,
      subFavoriteBerries: [ORAN, SITRUS],
      randomBonus: 'skill'
    };

    const event = expertModeEvent(settings);

    expect(event.name).toBe('Expert Mode Event');
    expect(event.modifiers).toHaveLength(3);

    // Test main berry frequency modifier
    expect(event.modifiers[0].type).toBe('PokemonInstance');
    expect(event.modifiers[0].leftValue).toBe('pokemon.frequency');
    expect(event.modifiers[0].rightValue).toBe(0.9);
    expect(event.modifiers[0].conditions?.[0].rightValue).toBe(settings.mainFavoriteBerry.name);

    // Test skill percentage boost (since randomBonus is 'skill')
    expect(event.modifiers[2].leftValue).toBe('pokemon.skillPercentage');
    expect(event.modifiers[2].rightValue).toBe(1.25);
  });

  it('should support mixed static and dynamic modifiers', () => {
    const mixedEventBuilder = EventBuilder<{ multiplier: number }>()
      .name('Mixed Event')
      .description('Combines static and dynamic modifiers')

      // Static modifier
      .addModifier({
        type: 'PokemonInstance',
        leftValue: 'pokemon.skillPercentage',
        operation: '*',
        rightValue: 1.1
      })

      // Dynamic modifier factory
      .addModifier((input) => ({
        type: 'PokemonInstance',
        leftValue: 'pokemon.frequency',
        operation: '*',
        rightValue: input.multiplier
      }));

    const event1 = mixedEventBuilder.build({ multiplier: 0.8 });
    expect(event1.modifiers).toHaveLength(2);
    expect(event1.modifiers[0].rightValue).toBe(1.1); // Static value
    expect(event1.modifiers[1].rightValue).toBe(0.8); // Dynamic value from input

    const mixedEventFunction = mixedEventBuilder.build();
    const event2 = mixedEventFunction({ multiplier: 0.9 });
    expect(event2.modifiers).toHaveLength(2);
    expect(event2.modifiers[0].rightValue).toBe(1.1); // Static value
    expect(event2.modifiers[1].rightValue).toBe(0.9); // Dynamic value from input
  });
});
