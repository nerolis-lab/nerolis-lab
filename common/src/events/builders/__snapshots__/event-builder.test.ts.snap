// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EventBuilder > should have proper type safety for modifiers 1`] = `
{
  "description": "A test event",
  "endDate": undefined,
  "modifiers": [
    {
      "leftValue": "pokemon.frequency",
      "operation": "*",
      "rightValue": 0.9,
      "type": "PokemonInstance",
    },
    {
      "leftValue": "pokemon.frequency",
      "operation": "*",
      "rightValue": 0.9,
      "type": "PokemonInstance",
    },
    {
      "leftValue": "pokemon.frequency",
      "operation": "*",
      "rightValue": "0.9",
      "type": "PokemonInstance",
    },
    {
      "leftValue": "pokemon.slkfjklsd",
      "operation": "*",
      "rightValue": 0.9,
      "type": "PokemonInstance",
    },
    {
      "conditions": [
        {
          "leftValue": "pokemon.berry.name",
          "operation": "=",
          "rightValue": "MAGO",
        },
      ],
      "leftValue": "pokemon.frequency",
      "operation": "*",
      "rightValue": 0.9,
      "type": "PokemonInstance",
    },
    {
      "conditions": [
        {
          "leftValue": "pokemon.berry.name",
          "operation": "=",
          "rightValue": "INVALID",
        },
      ],
      "leftValue": "pokemon.frequency",
      "operation": "*",
      "rightValue": 0.9,
      "type": "PokemonInstance",
    },
  ],
  "name": "Test Event",
  "startDate": undefined,
}
`;
